doc/man3/SSL_shutdown.pod

```diff
@@ -21,161 +21,242 @@ SSL_shutdown, SSL_shutdown_ex - shut down a TLS/SSL or QUIC connection
-it.
-
-The peer is still allowed to send data after receiving the close_notify
-event.
-When it is done sending data, it will send the close_notify alert.
-SSL_read() should be called until all data is received.
-SSL_read() will indicate the end of the peer data by returning <= 0
-and SSL_get_error() returning SSL_ERROR_ZERO_RETURN.
-
-=head2 Peer closes the connection
-
-If the peer already sent the close_notify alert B<and> it was
-already processed implicitly inside another function
-(L<SSL_read(3)>), the SSL_RECEIVED_SHUTDOWN flag is set.
-SSL_read() will return <= 0 in that case, and SSL_get_error() will return
-SSL_ERROR_ZERO_RETURN.
-SSL_shutdown() will send the close_notify alert, set the SSL_SENT_SHUTDOWN
-flag.
-If successful, SSL_shutdown() will return 1.
-
-Whether SSL_RECEIVED_SHUTDOWN is already set can be checked using the
-SSL_get_shutdown() (see also L<SSL_set_shutdown(3)> call.
+zero-initialised B<SSL_SHUTDOWN_EX_ARGS> structure. Currently, all extended
+arguments relate to usage with QUIC, therefore this call functions identically
+to SSL_shutdown() when not being used with QUIC.
+
+While the general operation of SSL_shutdown() is common between protocols, the
+exact nature of how a shutdown is performed depends on the underlying protocol
+being used. See the section below pertaining to each protocol for more
+information.
+
+In general, calling SSL_shutdown() in nonblocking mode will initiate the
+shutdown process and return 0 to indicate that the shutdown process has not yet
+completed. Once the shutdown process has completed, subsequent calls to
+SSL_shutdown() will return 1. See the RETURN VALUES section for more
+information.
+
+=head1 TLS AND DTLS-SPECIFIC CONSIDERATIONS
+
+Shutdown for SSL/TLS and DTLS is implemented in terms of the SSL/TLS/DTLS
+close_notify alert message. The shutdown process for SSL/TLS and DTLS
+consists of two steps:
+
+=over 4
+
+=item *
+
+A close_notify shutdown alert message is sent to the peer.
+
+=item *
+
+A close_notify shutdown alert message is received from the peer.
+
+=back
+
+These steps can occur in either order depending on whether the connection
+shutdown process was first initiated by the local application or by the peer.
+
+=head2 Locally-Initiated Shutdown
+
+Calling SSL_shutdown() on a SSL/TLS or DTLS SSL object initiates the shutdown
+process and causes OpenSSL to try to send a close_notify shutdown alert to the
+peer. The shutdown process will then be considered completed once the peer
+responds in turn with a close_notify shutdown alert message.
+
+Calling SSL_shutdown() only closes the write direction of the connection; the
+read direction is closed by the peer. Once SSL_shutdown() is called,
+L<SSL_write(3)> can no longer be used, but L<SSL_read(3)> may still be used
+until the peer decides to close the connection in turn. The peer might
+continuing sending data for some period of time before honouring the local
+application's shutdown request.
```

Comment:
I'm not sure it should be called a shutdown request, and so honouring also doesn't make sense to me.